### PR TITLE
feat: a new API design to allow customizations

### PR DESCRIPTION
**Pull request by lqhuang**
Hi,

I'm switching to very similar solutions like `jotai-ai` (ai sdk to unified api call + jotai for client-side state management) from `ai/rsc` recently.

Thanks for providing excellent reference implementations. But when try to use it in my case, I feel it's not that easy to adapt widely. I have implemented extended types based `CoreMessage` (which also could be seen as DB message or server state), but both original `useChat` and current `chatAtoms` are all difficult to customize client side message state, I almost need to write almost everything. Actually, the latest reference example [`ai-chatbot`](https://github.com/vercel/ai-chatbot) has been using `CoreMessage` for server state, `Message` (from `@ai/ui-utils`) for client state (which should be better to name as `UIMessage`). I hope the new design would achieve that easier. 

This PR implements some basic MVP to demonstrate my ideas:

1. Split main user interface to two functions `makeChatAtoms` and `useChat`. Let `useChat` expose useful and similar handlers/states like upstream `useChat` (`ai/react`), and left `makeChatAtoms` contains low level atoms for advanced users. Indeed, `makeChatAtoms` only controls how client interactive with remote stream data (append and receive new messages), and left all others into `useChat` in order that advanced users could implement theirs `inputAtoms`, `sumbitAtoms`. Built-in `useChat` could be just an official reference implementation has the same features with upstream projects.

2. Accept a user defined `messagesAtom` from outside. Like

```javascript
const messagesAtom = atom<Message[]>([]);

const {
  // state data containers,
  isLoadingAtom,
  errorAtom,
  dataAtom,

  // actions
  stopAtom,
  appendAtom,
  reloadAtom,

  // handlers
  onErrorAtom,
  onResponseAtom,
  onToolCallAtom,
  onFinishAtom,
} = makeChatAtoms({ messagesAtom });
``` 

In this way, users could decide how to prepare their UI messages, whatever derived from other states or initialed from fetch or even SSR.

```javascript
// Initial messages in SSR approach
// so `makeChatAtoms` doesn't need to consider 
// how to deal with initial messages or initial input.
import { useHydrateAtoms } from 'jotai/utils';

const messagesAtom = atom<Message[]>([]);

const Component = ({ initChatMesssages }:{ initChatMesssages: Message[] }) => {
  useHydrateAtoms([[messagesAtom, initChatMesssages]])

  const messages = useAtomValue(messagesAtom)

  return <Chat>{messages}<Chat/>
}
```

3. It would be easier to introduce more features and keep flexibility by using this architecture. We're able to add more states like `firstTokenReceived` or action like `deleteAtom` in the future without breaking up signatures of `useChat`.

Also, there are several issues I haven't confirmed and figured out clearly:

1. In current implements, all deprecated API and parameters are literally deprecated. And the `ai` project has already released alpha candidates of the next `v4` version.

2. Current callback handler like `onErrorAtom`, ..., `onResponseAtom` is initialed by `makeChatAtoms` and cloud be updated by `useChat`. I'm not sure whether it's a correct design or proper usage of atoms. For now, it's just implemented to fit upstream UI test cases easily.

3. I hope we could add a generic type interface for `Body` into `makeChatAtoms`. This data type can be shared between server endpoint and user client. It's more safe and ergonomic to make sure correct data being sent and parsed.

That's all. Thanks for your efforts!

Reject or feedback is all appreciate :). Have a nice day 🎉.

Sincerely,
Lanqing

### CODE REVIEW

**Comment by lqhuang** [comment]
@himself65 Could help me to check why test cases "setData" and "onFinish" cannot pass correctly.

I'm not an expert on React and jotai. My bad 🥲.

**Comment by himself65** [comment]
Thanks for your contribution! I will check

File: packages/jotai-ai/package.json [hunk 1]
```diff
@@ -30,18 +31,27 @@
   "author": "himself65 <himself65@outlook.com>",
   "license": "MIT",
   "devDependencies": {
+    "@ai-sdk/provider": "^0.0.26",
```
Comment (himself65) [inline packages/jotai-ai/package.json:34]: should be in `dependencies`

File: packages/jotai-ai/src/make-chat-atoms.ts [hunk 2]
```diff
@@ -0,0 +1,444 @@
...
+        ...metadata.headers,
+        ...chatRequest.headers,
+      },
+      credentials: metadata.credentials,
+      // handler
+      restoreMessagesOnFailure: () => undefined,
+      onResponse: response => get(onResponseAtom)?.(response),
+      onFinish: (message, options) => {
+        console.error('onFinish,', message, options);
+        console.error('onFinishAtom', typeof get(onFinishAtom));
```
Comment (himself65) [inline packages/jotai-ai/src/make-chat-atoms.ts:261]: no console?

File: packages/jotai-ai/src/make-chat-atoms.ts [hunk 3]
```diff
@@ -0,0 +1,444 @@
...
+        const onFinish = get(onFinishAtom);
+        if (onFinish) {
+          console.error('Error ????????');
```
Comment (himself65) [inline packages/jotai-ai/src/make-chat-atoms.ts:264]: remove unused log

**Review by himself65 — CHANGES_REQUESTED** [review]

File: packages/jotai-ai/src/make-chat-atoms.ts (hunk 3 above)
Comment (lqhuang) [reply packages/jotai-ai/src/make-chat-atoms.ts:264]: Oh. Sorry, I will remove them later

File: packages/jotai-ai/package.json (hunk 1 above)
Comment (lqhuang) [reply packages/jotai-ai/package.json:34]: Ok.

A small question, I just checked it that `@ai-sdk/provider` and `ai` currently is only used and imported in type definition (`import type ... from ...`). Do I still need to move to `dependencies`?

File: packages/jotai-ai/src/react/use-chat.ts [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+
+  // handlers
+  onErrorAtom,
+  onResponseAtom,
+  onToolCallAtom,
+  onFinishAtom,
+} = makeChatAtoms({ messagesAtom });
+
+export const useChat = (opts: UseChatOptions): UseChatReturn => {
+  const [input, setInput] = useAtom(inputAtom);
```
[the remark this answers is not in the document]
Comment (lqhuang) [reply packages/jotai-ai/src/react/use-chat.ts:166]: That's beyond my knowledge base. 😇😭🫠 (弱小而无助, 逃

**Review by himself65 — APPROVED** [review]
LGTM, what we can do more in the next steps:

- combine `makeChatAtoms` and `chatAtoms`
- align with vercel ai latest version
- add, update examples, docs

**Comment by lqhuang** [comment]
Love you 😍

**Comment by himself65** [comment]
I think ui test failed?

**Comment by lqhuang** [comment]
> I think ui test failed?

In where, CI or local env?

**Comment by lqhuang** [comment]
I see, you haven't fixed `setData` and `onFinish` cases yet.

> Could help me to check why test cases "setData" and "onFinish" cannot pass correctly.

Yeah, I can't understand why they failed before

**Pull request merged**